datafusion/core/src/physical_plan/aggregates/priority_queue.rs

```diff
@@ -0,0 +1,235 @@
+use uuid::Uuid;
+use std::collections::{BTreeMap, HashMap};
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use arrow::row::{OwnedRow, RowConverter, SortField};
+use arrow::util::pretty::print_batches;
+use arrow_array::{RecordBatch};
+use arrow_schema::{DataType, SchemaRef, SortOptions};
+use futures::stream::{Stream, StreamExt};
+use hashbrown::HashSet;
+use datafusion_common::DataFusionError;
+use datafusion_execution::TaskContext;
+use datafusion_physical_expr::{PhysicalExpr};
+use crate::physical_plan::aggregates::{aggregate_expressions, AggregateExec, evaluate_group_by, evaluate_many, group_schema, PhysicalGroupBy};
+use crate::physical_plan::{RecordBatchStream, SendableRecordBatchStream};
+use datafusion_common::Result;
+use datafusion_physical_expr::expressions::{Max, Min};
+
+pub(crate) struct GroupedPriorityQueueAggregateStream {
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    aggregate_arguments: Vec<Vec<Arc<dyn PhysicalExpr>>>,
+    group_by: PhysicalGroupBy,
+    group_converter: RowConverter,
+    value_converter: RowConverter, // TODO: use accumulators
+    group_to_val: HashMap<OwnedRow, OwnedRow>, // TODO: BTreeMap->BinaryHeap, OwnedRow->Rows
+    val_to_group: BTreeMap<OwnedRow, HashSet<OwnedRow>>,
```

Comment:
This probably deserves some attention as well - creating hashset per value and doing some operations on them probably will be slow.

Reply:
BinaryHeap sounds good

Reply:
Probably better for a follow up PR, but I think we could also specialize for very small limits, e.g. `limit 1` doesn't have to use a heap but can just track one max/min, small `n` (maybe up to 30) probably can sort or use a sorting network, etc.

Reply:
> BinaryHeap sounds good

It does, but the standard library implementation doesn't have all the requisite [traversal and rewrite methods](https://github.com/avantgardnerio/arrow-datafusion/blob/c6ce44f3bc5b787c96c52e574727bfd1f820e05d/datafusion/core/src/physical_plan/aggregates/priority_queue.rs#L260), which was why I started with `TreeMap`.

I could see using something like https://lib.rs/crates/btreemultimap, but most of the multimaps I can find are just wrappers around `BTreeMap` with heap allocation of a `Vec` or `Set` similar to what I had already.

I can begin hunting for a suitable off-the-shelf implementation, or build my own, but advice from those here would be appreciated.

Reply:
Oh, I guess those are heap properties, not implementation properties 😊 . 

It looks like there is an `.iter().find()` (`O(n)`) and `.retain()` (`O(n)`). A worst-case scenario sounds horrible (limit*row_count), but assuming a uniform distribution of data it would much more likely be logarithmic in practice. Maybe we're better off with the BTree for all around performance? Just need to find or make one that avoids or minimizes `malloc`s?

Reply:
@alamb I think the `BTreeMap<OwnedRow, HashSet<OwnedRow>>` might be the best the most efficient overall implementation.

I'm working on a version that:
1. doesn't modify anything if the new value seen is less than the min (or greater than max) of the working set
2. doesn't modify anything if the value for a group_id meets # 1
3. adds the HashSets if needed
4. upon filling all slots with HashSets, re-uses them to avoid any further heap allocations

@Dandandan hopefully # 4 in that list satisfies your previous concern? Taking it to it's logical conclusion one could envision fixed size HashSets up front (https://github.com/Simsys/fchashmap) but then we'd be talking limit^2 memory.

Reply:
Sure `HashSet` is good as first start. 

Some ideas on this:

* HashSet might use the `limit` value for initial capacity (for small limits)
* instead of `HashSet` use `RawTable` and store rows seperately
* It's possible for really small `n` or `n=1` we could use an Vec or singular value (max/min for the group) for n=1

Reply:
@Dandandan I can keep a [pool of them](https://github.com/avantgardnerio/arrow-datafusion/blob/6a535213e2a4ae8ef00f1b541a2f569bc5e3a2b7/datafusion/core/src/physical_plan/aggregates/priority_queue.rs#L182), and [pre-allocate](https://github.com/avantgardnerio/arrow-datafusion/blob/6a535213e2a4ae8ef00f1b541a2f569bc5e3a2b7/datafusion/core/src/physical_plan/aggregates/priority_queue.rs#L192), but the code [gets gnarly fast](https://github.com/avantgardnerio/arrow-datafusion/blob/6a535213e2a4ae8ef00f1b541a2f569bc5e3a2b7/datafusion/core/src/physical_plan/aggregates/priority_queue.rs#L196).

Perhaps we just go with the simple (and malloc heavy) code to start, then put it through a perf test & profiler and go from there?

Reply:
I like @thinkharderdev 's suggestion of binary search on a Vec. It's still log(n), but insertion and deletion will do a lot of shifting things around, but the things are 10 byte OwnedRows, so 1000 byte `memcpy`s would play pretty well with cache pipelines and might be our best option.

Reply:
Yeah of course :).

I think once we have some benchmarks / queries to test against we can do some profiling and see if it's worth optimizing.

Reply:
Binary search (essentially a heap) is fine.

Reply:
I swapped out the `BTreeSet` for binary heap because that allowed me to decouple the types and support all permutations of `<K,V>`, while keeping approximately the same level of performance.